### PR TITLE
IDENTITY-5577: Fixing caching issue in Policy Combining Algo update

### PR DESCRIPTION
**Pull request by dimuthud**

### CODE REVIEW

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java [hunk 1]
```diff
@@ -138,30 +138,34 @@ public void invalidateCache() {
...
+            log.debug("Check the invalidation state of all cache, isCacheInvalid: " + isInvalid);
         }
 
+        return isInvalid;
     }
 
+    public void resetCacheInvalidateState() {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+
```
Comment (mefarazath) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java]: unnecessary new line.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/store/DefaultPolicyDataStore.java [hunk 2]
```diff
@@ -117,6 +118,10 @@ public void setGlobalPolicyAlgorithm(String policyCombiningAlgorithm) throws Ent
             policyCollection.setMediaType(PDPConstants.REGISTRY_MEDIA_TYPE);
             policyCollection.setProperty("globalPolicyCombiningAlgorithm", policyCombiningAlgorithm);
             registry.put(policyDataCollection, policyCollection);
+
+            // performing cache invalidation
+            EntitlementEngine.getInstance().getPolicyCache().invalidateCache();
+
```
Comment (mefarazath) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/store/DefaultPolicyDataStore.java]: shall we remove this line?

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java [hunk 3]
```diff
@@ -138,30 +138,34 @@ public void invalidateCache() {
...
-        }
-        if(log.isDebugEnabled()){
-            log.debug("Check the invalidation state of all cache : " + state);
         }
 
-        if(state==1){
-            return true ;
-        }else{
-            return false ;
+        boolean isInvalid = state == 1;
```
Comment (ruwanta) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java]: For better readability please change, 
boolean isInvalid = state == 1;
-- > boolean isInvalid = (state == INVALID_STATE);
and have a constant for INVALID_STATE

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/finder/CarbonPolicyFinder.java [hunk 4]
```diff
@@ -223,6 +223,7 @@ public PolicyFinderResult findPolicy(EvaluationCtx context) {
             init(this.finder);
             policyReferenceCache.clear();
             EntitlementEngine.getInstance().clearDecisionCache();
+            EntitlementEngine.getInstance().getPolicyCache().resetCacheInvalidateState();
```
Comment (ruwanta) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/finder/CarbonPolicyFinder.java]: What is the guarantee that EntitlementEngine.getInstance().getPolicyCache() is not null always? 
Better option is to add a method to EntitlementEngine.class
public void clearPolicyCacheForCurrentTenent() {
 if(getPolicyCache() != null) {
getPolicyCache(). resetCacheInvalidateState();
}

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java (hunk 1 above)
Comment (dimuthud) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java]: Done.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/store/DefaultPolicyDataStore.java (hunk 2 above)
Comment (dimuthud) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/store/DefaultPolicyDataStore.java]: Done.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java (hunk 3 above)
Comment (dimuthud) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/cache/PolicyCache.java]: Done.

File: components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/finder/CarbonPolicyFinder.java (hunk 4 above)
Comment (dimuthud) [inline components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/policy/finder/CarbonPolicyFinder.java]: This is a used in Single design pattern. Hence this wont be null.

**Pull request merged**